### PR TITLE
improve softmax implementation

### PR DESCRIPTION
**Pull request by bobchennan**
Instead of taking maximal of whole tensor, it is better to get maximal for each vector where softmax is calculated.

### CODE REVIEW

**Comment by taehoonlee** [comment]
Thank you for the PR @bobchennan. Would you like to explain the reason you added `and x.ndim == 2`? Is there any issue in `T.nnet.softmax(x)` where `x.ndim > 2`?

**Comment by bobchennan** [comment]
@taehoonlee Yes according to the [official document](http://deeplearning.net/software/theano/library/tensor/nnet/nnet.html#theano.tensor.nnet.nnet.softmax), theano only handle 2-dimensional case.

**Comment by bobchennan** [comment]
Or you can look at their code:
https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/nnet.py#L429

**Comment by taehoonlee** [comment]
LGTM. Please add a unit test for 3D cases to `tests/keras/activations_test.py`. Additionally, `softmax` in `keras/activations.py` might be cleaned up.

**Comment by taehoonlee** [comment]
@bobchennan, I don't know the original intention. I think it is better to remain the 1D case returning ValueError, and focus the 3D case. Can you fix the travis errors? The input array can not be fed into placeholder because the `get_standard_values` produces a 2D array (see [here](https://travis-ci.org/keras-team/keras/jobs/372101595#L2288)).

**Comment by bobchennan** [comment]
Seems to be a bug and I submitted to #10073

File: keras/activations.py [hunk 1]
```diff
@@ -20,12 +20,11 @@ def softmax(x, axis=-1):
 
     # Returns
         Tensor, output of softmax transformation.
-
-    # Raises
-        ValueError: In case `dim(x) == 1`.
```
Comment (fchollet) [inline keras/activations.py]: Do not remove this section from the docstring (the exception is still there).

**Comment by gabrieldemarmiesse** [comment]
Wow my bad. I tried to push a commit to this branch, but it didn't go as expected.

**Comment by gabrieldemarmiesse** [comment]
I guess it still works out in the end but the history won't be clean at all. And also the "files changed" section is unreadable. If anyone knows how to fix this git mess, thanks. And really sorry for that.

**Comment by gabrieldemarmiesse** [comment]
Closing because #11189 was merged.

**Pull request closed without merging**